src/Microsoft.FeatureManagement/VariantServiceProvider.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+//
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.FeatureManagement
+{
+    /// <summary>
+    /// Used to get different implementations of TService depending on the assigned variant from a specific variant feature flag.
+    /// </summary>
+    internal class VariantServiceProvider<TService> : IVariantServiceProvider<TService> where TService : class
+    {
+        private readonly IEnumerable<TService> _services;
+        private readonly IVariantFeatureManager _featureManager;
+        private readonly string _variantFeatureName;
+        private readonly ConcurrentDictionary<string, TService> _variantServiceCache;
+
+        /// <summary>
+        /// Creates a variant service provider.
+        /// </summary>
+        /// <param name="services">Implementation variants of TService.</param>
+        /// <param name="featureManager">Feature manager to get the assigned variant of the variant feature flag.</param>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="services"/> is null.</exception>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="featureManager"/> is null.</exception>
+        public VariantServiceProvider(IEnumerable<TService> services, IVariantFeatureManager featureManager)
+        {
+            _services = services ?? throw new ArgumentNullException(nameof(services));
+            _featureManager = featureManager ?? throw new ArgumentNullException(nameof(featureManager));
+            _variantServiceCache = new ConcurrentDictionary<string, TService>();
+        }
+
+        /// <summary>
+        /// The variant feature flag used to assign variants.
+        /// </summary>
+        public string VariantFeatureName
+        {
+            get => _variantFeatureName;
+
+            init
+            {
+                _variantFeatureName = value ?? throw new ArgumentNullException(nameof(value));
```

Comment:
> It's a required for the successful usage of the object isn't it?

Yes.
I used to use the feature flag name as a constructor parameter in this commit: 176cf78616366b29309c80c3cf1dad9ddb18b3e6

But then I decide to put it in a public property, because in this way, which variant feature flag is bound to this variant service provider can be known from the outside. 
I think the interface type and the feature flag name are equal in status. We can know the interface type from the type `VariantServiceProvider<TService>` but we won't know the feature flag name if we put it in a private property.

I know that it is unnecessary to expose the feature flag name to public at this moment , the `VariantServiceProvider` class is actually internal. So I won't mind putting the feature flag name in the constructor parameter.

Reply:
Please put it in the constructor. It's especially unexpected that `GetServiceAsync` asserts that feature name is not null when nothing about the instantiation of the type enforces it to be non-null.

Reply:
`Debug.Assert` should only be used to assert conditions that are validated by public entry point and a reader can verify to be valid.